unixfs/io/dagreader.go

```diff
@@ -19,14 +20,29 @@ var ErrIsDir = errors.New("this dag node is a directory")
 type DagReader struct {
 	serv         mdag.DAGService
 	node         *mdag.Node
-	buf          io.Reader
-	fetchChan    <-chan *mdag.Node
+	pbdata       *ftpb.Data
+	buf          ReadSeekCloser
+	promises     []mdag.NodeGetter
 	linkPosition int
+	offset       int64
+
+	// Our context
+	ctx context.Context
+
+	// Context for children
+	fctx   context.Context
+	cancel func()
```

Comment:
- `fctx` isn't set anywhere.
- `cancel` corresponds to `ctx`
- when you want to embed the ctx as state in an object, with cancellation as the main purpose, try `goprocess`: 

```go
type DagReader struct {
  ...
  proc goprocess.Process
}

func (dr *DagReader) Close() error {
  return dr.proc.Close()
}
```